llvm/test/Analysis/ValueTracking/phi-known-bits.ll

```diff
@@ -424,3 +424,330 @@ T:
 F:
   br label %T
 }
+
+define i1 @phi_sub_recursiveGEP_knownNonZero1(ptr noundef %val1) {
+; CHECK-LABEL: @phi_sub_recursiveGEP_knownNonZero1(
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    [[CMP_I:%.*]] = icmp eq ptr [[VAL1:%.*]], null
+; CHECK-NEXT:    br i1 [[CMP_I]], label [[_Z9STRINGLENPKS_EXIT:%.*]], label [[WHILE_COND_I:%.*]]
+; CHECK:       while.cond.i:
+; CHECK-NEXT:    [[A_PN_I:%.*]] = phi ptr [ [[TEST_0_I:%.*]], [[WHILE_COND_I]] ], [ [[VAL1]], [[ENTRY:%.*]] ]
+; CHECK-NEXT:    [[TEST_0_I]] = getelementptr inbounds i8, ptr [[A_PN_I]], i64 1
+; CHECK-NEXT:    [[TMP0:%.*]] = load i8, ptr [[TEST_0_I]], align 2
+; CHECK-NEXT:    [[CMP3_NOT_I:%.*]] = icmp eq i8 [[TMP0]], 0
+; CHECK-NEXT:    br i1 [[CMP3_NOT_I]], label [[WHILE_END_I:%.*]], label [[WHILE_COND_I]]
+; CHECK:       while.end.i:
+; CHECK-NEXT:    br label [[_Z9STRINGLENPKS_EXIT]]
+; CHECK:       _Z9stringlenPKs.exit:
+; CHECK-NEXT:    ret i1 [[CMP_I]]
+;
+entry:
+  %cmp.i = icmp eq ptr %val1, null
+  br i1 %cmp.i, label %_Z9stringlenPKs.exit, label %while.cond.i
+
+while.cond.i:
+  %a.pn.i = phi ptr [ %test.0.i, %while.cond.i ], [ %val1, %entry ]
+  %test.0.i = getelementptr inbounds i8, ptr %a.pn.i, i64 1
+  %0 = load i8, ptr %test.0.i, align 2
+  %cmp3.not.i = icmp eq i8 %0, 0
+  br i1 %cmp3.not.i, label %while.end.i, label %while.cond.i
+
+while.end.i:
+  %sub.ptr.lhs.cast.i = ptrtoint ptr %test.0.i to i64
+  %sub.ptr.rhs.cast.i = ptrtoint ptr %val1 to i64
+  %sub.ptr.sub.i = sub i64 %sub.ptr.lhs.cast.i, %sub.ptr.rhs.cast.i
+  br label %_Z9stringlenPKs.exit
+
+_Z9stringlenPKs.exit:
+  %retval.0.i = phi i64 [ %sub.ptr.sub.i, %while.end.i ], [ 0, %entry ]
+  %bool = icmp eq i64 %retval.0.i, 0
```

Comment:
This code will get converted into something like this by InstCombine:
```suggestion
  %x = icmp eq i64 %sub.ptr.lhs.cast.i, %sub.ptr.rhs.cast.i
  br label %_Z9stringlenPKs.exit

_Z9stringlenPKs.exit:
  %bool = phi i1 [ %x, %while.end.i ], [ true, %entry ]
```
At which point the sub of ptrtoint pattern no longer exists.

So I don't understand how your code even does anything here -- unless you are relying on your fold triggering before this canonicalization, which is obviously not something you can do.

Reply:
Thanks for the review. 
For the given example code visitPHINode() should be invoked first and the code then identifies it as a knownNonZero() scenario and does the transformation. Exclusively this can be a check done in visitIcmp() too before we fold the icmp(sub) which could help fold additional scenarios.
Additionally since we are identifying it is a isNonZeroSub() this can be applied beyond scenarios where the only use of phi is a icmp. This can also be applied to 
1. Multiple uses of phi with or, icmp
2. Single use of phi with or(icmp)
Some of these examples can be added for the patch.

Reply:
Okay, so you *are* relying on transform order. In that case, what you need to do is first implement the fold for canonical IR (that is, for the case where you have a `icmp eq i64 %sub.ptr.lhs.cast.i, %sub.ptr.rhs.cast.i` style comparison). Once that has landed, we can go back and consider whether it is necessary or worthwhile to handle the non-canonical form.

Reply:
Ok so in this patch we are handling the sub of 2 ptr2ints.
But I am not specifically relying on the transform order for this pattern. Sure in the example provided the phi in InstCombine hits first. However there is use to the isNonZeroSub() from InstCombineCompares as well. Which means for the scneario where we have the below
```
  %sub.ptr.lhs.cast.i = ptrtoint ptr %test.0.i to i64
  %sub.ptr.rhs.cast.i = ptrtoint ptr %val1 to i64
  %sub.ptr.sub.i = sub i64 %sub.ptr.lhs.cast.i, %sub.ptr.rhs.cast.i
  %bool = icmp eq i64 %sub.ptr.sub.i, 0
  br label %_Z9stringlenPKs.exit

_Z9stringlenPKs.exit:
  %retval.0.i = phi i1 [ %bool, %while.end.i ], [ false, %entry ]
```

The same does get folded away as well. As I specified maybe adding some test cases showing this would give more clarity.
So basically irrespective of Transform order if we have a sub(pt2int, ptr2int) it should be folded away.

Reply:
Updated the PR with test showing folds with icmp(sub(ptr2int,ptr2int), 0) and scenarios where it wont fold from Compares(multi use phi, single use with or(icmp))

Reply:
This doesn't really address my concern. My problem here is that you are trying to handle a complex pattern, while there exists a simpler pattern that is not being handled yet. You need to handle the simple pattern *first*, and the complex pattern later.

Reply:
Thanks. That makes sense. I have done this in the PR https://github.com/llvm/llvm-project/pull/70459.